### PR TITLE
Add UDP input back

### PR DESCRIPTION
**Pull request by renan-**
Hi there,

Seeing that my previous pull request is not likely to get merged because of the big refactoring going on, here's a PR to add UDP support back (not sure why it was removed) in alpha1. I hope this will be helpful, and I'm looking forward to the next release ! (got rc31 running in production and still seeing some issues as well as feeling the lack of CQs).

Keep up the good work,
Renan.

- [x] CHANGELOG.md updated
- [x] Rebased/mergable
- [x] Tests pass
- [x] Sign [CLA](http://influxdb.com/community/cla.html) (if not already signed)

### CODE REVIEW

**Comment by corylanou** [comment]
@renan- Thanks for the work.  The reason we removed it was we are not going to support json on this endpoint.  We are only going to support the new line protocol here likely.  I'll get @pauldix to confirm though to be sure.  Regardless, I imagine at the very least we would do something similar to the http endpoint and detect json vs. line protocol to support both.

**Comment by pauldix** [comment]
Yeah, I'd prefer to have UDP support only the line protocol. Which would mean it would need configuration options like the Graphite and CollectD input plugins. Set which database and retention policy you're writing into.

See #2696 for discussion on the merits of line protocol over JSON.

**Comment by renan-** [comment]
Thanks for the quick answers guys ! I desperately need UDP support for my use case (to avoid any latency when writing data) so would you be interested in another PR supporting line protocol instead of JSON ?

**Comment by pauldix** [comment]
@renan- that would be great, we'd gladly take it. Going to cut RC32 tomorrow so if you can get it in time, it'll be included! Thanks again for helping out!

**Comment by pauldix** [comment]
@renan- oh, another thing to look at is to have it use the batcher that graphite and collectd now use.

**Comment by renan-** [comment]
Thanks for the tip, starting to work on it.

**Comment by renan-** [comment]
Updated the code, now waiting for your feedback.

File: cmd/influxd/run/config.go [hunk 1]
```diff
@@ -77,6 +78,7 @@ type Config struct {
 	Graphites []graphite.Config `toml:"graphite"`
 	Collectd  collectd.Config   `toml:"collectd"`
 	OpenTSDB  opentsdb.Config   `toml:"opentsdb"`
+	UDP       udp.Config        `toml:"udp"`
```
Comment (corylanou) [inline cmd/influxd/run/config.go]: please add a test in `config_test.go` for this addition of the config.

File: services/udp/config.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+type Config struct {
+	Enabled    bool   `toml:"enabled"`
+	BindAdress string `toml:"bind-adress"`
+	Port       int    `toml:"port"`
+
+	Database     string        `toml:"database"`
+	BatchSize    int           `toml:"batch-size"`
+	BatchTimeout toml.Duration `toml:"batch-timeout"`
+}
```
Comment (corylanou) [inline services/udp/config.go]: please add associated tests to `config_test.go`

File: services/udp/service.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	iface string
+}
+
+func NewService(c Config) *Service {
+	server := NewServer(c.Database)
+	server.SetBatchSize(c.BatchSize)
+	server.SetBatchTimeout(time.Duration(c.BatchTimeout))
+
+	return &Service{
+		iface:  net.JoinHostPort(c.BindAdress, strconv.Itoa(c.Port)),
```
Comment (corylanou) [inline services/udp/service.go]: We are moving to a `addr` variable name here.  Also, we no longer use port as part of the config.  `BindAddress` from the config should now be in the format of `host:port`.

File: services/udp/config.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package udp
+
+import "github.com/influxdb/influxdb/toml"
+
+type Config struct {
+	Enabled    bool   `toml:"enabled"`
+	BindAdress string `toml:"bind-adress"`
+	Port       int    `toml:"port"`
```
Comment (corylanou) [inline services/udp/config.go]: No longer using `Port` in our configs.  BindAddress should now be in the format of `host:port`

File: etc/config.sample.toml [hunk 5]
```diff
@@ -73,6 +73,9 @@ enabled = false
 enabled = false
 # bind-address = "0.0.0.0"
 # port = 4444
```
Comment (corylanou) [inline etc/config.sample.toml]: Remove, no longer using port.

File: services/udp/udp.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		s.Logger.Printf("Failed to resolve UDP address %s: %s", iface, err)
+		return err
+	}
+
+	s.conn, err = net.ListenUDP("udp", s.addr)
+	if err != nil {
+		s.Logger.Printf("Failed to set up UDP listener at address %s: %s", s.addr, err)
+		return err
+	}
+
```
Comment (corylanou) [inline services/udp/udp.go]: We need to check to see if they database was set in the config, if not, throw an error.  See for reference: https://github.com/influxdb/influxdb/blob/master/services/collectd/collectd.go#L60

File: services/udp/udp.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	s.Logger.Printf("Started listening on %s", iface)
+
+	batcher := tsdb.NewPointBatcher(s.batchSize, s.batchTimeout)
+	batcher.Start()
+
+	// Start processing batches.
+	var wg sync.WaitGroup
+	done := make(chan struct{})
+	wg.Add(1)
+	go func() {
```
Comment (corylanou) [inline services/udp/udp.go]: Can you make this it's own method instead of an anonymous function.

File: services/udp/udp.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+					s.Logger.Printf("Wrote a batch of %d points to %s", len(batch), s.database)
+				}
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	s.wg.Add(1)
+	go func() {
```
Comment (corylanou) [inline services/udp/udp.go]: Please make a method for this.

File: services/udp/udp.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	s.Logger.Printf("Started listening on %s", iface)
+
+	batcher := tsdb.NewPointBatcher(s.batchSize, s.batchTimeout)
+	batcher.Start()
+
+	// Start processing batches.
+	var wg sync.WaitGroup
+	done := make(chan struct{})
```
Comment (corylanou) [inline services/udp/udp.go]: This needs to be part of the server struct. Also, I don't see us actually `close` the channel, which needs to be done in `server.Close`

File: services/udp/udp.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+	return nil
+}
+
+func (s *Server) Close() error {
+	var err error
+	if s.conn != nil {
+		err = s.conn.Close()
+	}
+	s.wg.Done()
```
Comment (corylanou) [inline services/udp/udp.go]: I think this should be a `s.wg.Wait()`

File: services/udp/udp.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+	go func() {
+		defer s.wg.Done()
+		for {
+			buf := make([]byte, udpBufferSize)
+			n, _, err := s.conn.ReadFromUDP(buf)
+			if err != nil {
+				s.Logger.Printf("Failed to read UDP message: %s", err)
+				batcher.Flush()
+				close(done)
+				wg.Wait()
```
Comment (corylanou) [inline services/udp/udp.go]: I think this should be a `s.wg.Done()`

**Comment by corylanou** [comment]
@renan- Thanks for the hard work!  Certainly headed in the correct direction.  Please let me know if you have any questions.  I know not all of our services are in perfect shape yet either, but feel free to look at collectd to get an idea of some of the patterns we are currently using (I reference some as well in my comments).

**Comment by renan-** [comment]
@corylanou You're welcome! I updated the PR once again based on your comments. I still wonder if the retention policy should be in config or always set to the default one. One more thing: what about ConsistencyLevel value when creating a WritePointsRequest ?

File: services/udp/udp.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+	defer s.wg.Done()
+
+	s.batcher.Start()
+	for {
+		buf := make([]byte, UDPBufferSize)
+		n, _, err := s.conn.ReadFromUDP(buf)
+		if err != nil {
+			s.Logger.Printf("Failed to read UDP message: %s", err)
+			s.batcher.Flush()
+			return
```
Comment (renan-) [inline services/udp/udp.go]: @corylanou Should I continue instead (just like in collectd) ? And what about the call to batcher.Flush() ?

File: cmd/influxd/run/server_test.go [hunk 13]
```diff
@@ -1010,6 +1010,9 @@ func NewConfig() *run.Config {
 	c.HTTPD.Enabled = true
 	c.HTTPD.BindAddress = "127.0.0.1:0"
 	c.HTTPD.LogEnabled = testing.Verbose()
+
+	c.UDP.BindAddress = "127.0.0.1:0"
+	c.UDP.Database = "db0"
```
Comment (corylanou) [inline cmd/influxd/run/server_test.go]: We actually don't want to add these.  This would enable the UDP endpoint for every test.  We use this function to create a `base` config for the tests to use, and then turn on specific endpoints in the tests themselves.  Simply revert the change to this file and we should be fine.

**Comment by corylanou** [comment]
@renan- Awesome!  Thanks for all the work.  Very much appreciated!  Had one last minor comment.  Looks like something changed that is preventing a merge so you'll have to rebase quick to make this mergeable.

For now let the default RP be used.  It can be a feature we add in the future to specify an RP.  If we do, all endpoints should likely support it as well.

Same with consistency level.  feels like that should be a config option as well, but I'm fine with this PR not addressing it. Can be an easy feature to add.

Thanks again for the quick turn around.  Hopefully we'll get this merged for tomorrow's RC.

File: cmd/influxd/run/server_test.go (hunk 13 above)
Comment (renan-) [inline cmd/influxd/run/server_test.go]: Well if I revert this change, tests still run and fail beacause there is no db nor bindaddress set

Comment (renan-) [inline cmd/influxd/run/server_test.go]: Tests on CircleCI won't pass without these

Comment (corylanou) [inline cmd/influxd/run/server_test.go]: I think the real problem here is that the server isn't observing if UDP is enabled, and actually adding it to the services.  I'm guessing a couple of our endpoints have that problem.  I'll merge this, and then do another PR that fixes services starting that aren't enabled.
